### ISSUE TITLE
Appveyor

### ISSUE THREAD

**Issue by MSP-Greg** [bundler/bundler#6001]
@segiddins

```
rake spec
2463 examples, 1157 failures, 1 pending  03Sept 1
2463 examples,  664 failures, 3 pending  04Sept 1
2463 examples,  478 failures, 3 pending  04Sept 2
2307 examples,  287 failures, 1 pending  11Sept
```

I saw Travis took a while, so I started the tests and returned later for the output.  Obviously, some major things are wrong.

Since MinGW / Windows is new, figured I might as well go for maximum pain, using
```
ruby 2.5.0dev (2017-09-03 trunk 59733) [x64-mingw32]
```

First goal was just to get spec to run.  A few things:

1. ruby-graphviz gem installed, but it seems that Graphviz itself is not a good match for Windows.  I'm not even sure how its used in testing.

2. Most ruby win installs use rb-readline, and, on first pass, it didn't it play well with the testing (CLI?).  I believe my builds are the only current MinGW or mswin builds that also compile readline (rb-readline overrides readline.so).  I needed to bypass rb-readline and use the readline.so version for spec to run.

3. Git - I did some PR's in Core and rubygems that pertained to using Git in testing.  Standard windows install places Git in a path with a space.  I've got a symlink without a space, but I didn't have that in my path for the tests. Some test failures/errors seemed to be Git related...

4. ENV['HOME'] - some error messages, on many windows installs, there's a space in this path.  I replaced with e:/temp for the test run.

This might take a while.  I think (amongst the many things he's working on) that @hsbt might also be working on this for ruby core...

**Comment by segiddins** [comment]
> ruby-graphviz gem installed, but it seems that Graphviz itself is not a good match for Windows. I'm not even sure how its used in testing.

I'm fine skipping those specs on windows, `bundle viz` is being removed from Bundler core in 2.0 anyways

> Most ruby win installs use rb-readline, and, on first pass, it didn't it play well with the testing (CLI?). I believe my builds are the only current MinGW or mswin builds that also compile readline (rb-readline overrides readline.so). I needed to bypass rb-readline and use the readline.so version for spec to run.

I'm not quite sure why this is causing an issue? We don't (directly) use readline anywhere

> Git - I did some PR's in Core and rubygems that pertained to using Git in testing. Standard windows install places Git in a path with a space. I've got a symlink without a space, but I didn't have that in my path for the tests. Some test failures/errors seemed to be Git related...

cleaning up the tests to avoid a shell has been on my TODO list for a while. But to the best of my knowledge, nothing actually uses the full path to `git` and just relies on backtick & the shell finding it properly in `PATH`

> ENV['HOME'] - some error messages, on many windows installs, there's a space in this path. I replaced with e:/temp for the test run.

we're not going to fix this one any time soon, but the specs should be failing fast (intentionally) when run from a path with spaces (or other weird characters, for that matter)

**Comment by MSP-Greg** [comment]
> [Graphviz]  I'm fine skipping those specs on windows

Right now I'm just running `rake spec`.  and I don't recall a task with them.  Are they part of the normal tests?

>[readline] I'm not quite sure why this is causing an issue? We don't (directly) use readline anywhere

Thought I came across some requires.  Maybe something different than the std-lib readline?  I check again, as things are more stable now, but with the rb-readline version, the testing stopped / froze.  Could one of the gems be using for mocking a CLI?

Re git, you may be correct.  I did fix some backtick statements that use single quotes.  Windows 'prefers' double quotes.

Re the path issue, I think I've got it worked out.

Question - I have around 100 failures with the following:
```
E:/ruby25_09-03/lib/ruby/2.5.0/rubygems/remote_fetcher.rb:350:in `correct_for_windows_path'
E:/ruby25_09-03/lib/ruby/2.5.0/rubygems/remote_fetcher.rb:243:in `fetch_file'
E:/ruby25_09-03/lib/ruby/2.5.0/rubygems/remote_fetcher.rb:292:in `fetch_path'
E:/GitHub/bundler/lib/bundler/rubygems_integration.rb:744:in `fetch_specs'
E:/GitHub/bundler/lib/bundler/rubygems_integration.rb:754:in `fetch_all_remote_specs'
E:/GitHub/bundler/lib/bundler/fetcher/index.rb:10:in `specs'
E:/GitHub/bundler/lib/bundler/fetcher.rb:129:in `block in specs'
-- stopped here --
```
I think it's because of a URI::Generic (file:// scheme) issue.  Some begin with `file://E/path_to_whatever`, others are `file:E:/path_to_whatever`.  Obviously a windows problem.  My question - given that this crosses the Bundler / RubyGems boundary, am I for looking where the problem begins (the bad URI), or where it causes an error?  My guess is the first, but I thought I'd ask.  I need to spend some more time with the Bundler and RubyGems code/docs, as I really don't have a big picture view yet, especially where the divide is...

Lastly, I haven't gotten deep into the testing, I'm just running the tests to a log, and checking how many hits I get on various backtrace items.

**Comment by segiddins** [comment]
> Are they part of the normal tests?

yup, `spec/commands/viz_spec.rb`

> Thought I came across some requires

only thing I found is coming from https://github.com/erikhuda/thor

> `correct_for_windows_path`

my guess is we can just fix the specs to be consistent with the right number of slashes, but rubygems should probably handle an arbitrary number of preceding slashes, as I believe that's technically allowed in URLs?

**Comment by MSP-Greg** [comment]
> only thing I found is coming from https://github.com/erikhuda/thor

I didn't log where the tests stopped.  I'll do so and investigate.

> my guess is we can just fix the specs to be consistent with the right number of slashes, but rubygems should probably handle an arbitrary number of preceding slashes, as I believe that's technically allowed in URLs?

Re fixing the spec, I don't want to break Travis testing, and I knwo very little about *nix/OSX.  Don't know anything about URL's/URI's specs.  Regardless, I'll get it patched, we can figure out where to put it when I get the failures down to a reasonable number.

Thanks.  I'll keep updating the first message with my current failure count, and pop in if I've got questions...

**Comment by MSP-Greg** [comment]
@hsbt

Given that you seem to be leading the trunk/core task of 'gem related' issues, do you have a preference for whether I work on bundler appveyor/windows testing here or in ruby/ruby?

I suspect the task is on your (and/or others) tasklist, but I have no idea whether you've started, and if so, how far along you are.

The things I've found are typical windows issues.  Most of my time is spent tracking down where the 'origin' of the problems are.  At present, Windows, trunk, and bundler don't play well together, so I'm also learning/refreshing my knowledge of bundler & rubygems...

I'm kind of assuming that mswin & MinGW share many of the same issues, but since MinGW is built in a kind of 'half & half' environment, the issues may not be identical.  For that reason, I started here using my MinGW trunk builds.

**Comment by colby-swandale** [comment]
We've merged #6899 as an alternative to Appveyor
